File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/redis/SurrenderVoteStore.kt
package io.github.kapu.turtlesoup.redis

import io.github.kapu.turtlesoup.config.JsonConfig
import io.github.kapu.turtlesoup.config.RedisConstants
import io.github.kapu.turtlesoup.config.RedisKeys
import io.github.kapu.turtlesoup.models.SurrenderVote
import io.github.kapu.turtlesoup.utils.RedisException
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import org.redisson.api.RedissonReactiveClient
import java.time.Duration
import org.redisson.client.RedisException as RedissonException

/** 항복 투표 저장소 (Reactive) */
class SurrenderVoteStore(
    private val redisson: RedissonReactiveClient,
) {
    private val json = JsonConfig.lenient

    suspend fun get(chatId: String): SurrenderVote? {
        val key = key(chatId)
        return try {
            val bucket = redisson.getBucket<String>(key)
            val value = bucket.get().awaitSingleOrNull() ?: return null
            json.decodeFromString<SurrenderVote>(value)
        } catch (e: RedissonException) {
            logger.error(e) { "vote_load_failed chat_id=$chatId" }
            throw RedisException("Failed to load vote: $chatId", e)
        } catch (e: SerializationException) {
            logger.error(e) { "vote_load_failed chat_id=$chatId" }
            throw RedisException("Failed to load vote: $chatId", e)
        }
    }

    suspend fun save(
        chatId: String,
        vote: SurrenderVote,
    ) {
        val key = key(chatId)
        try {
            val value = json.encodeToString(vote)
            val bucket = redisson.getBucket<String>(key)
            bucket.set(value, Duration.ofSeconds(RedisConstants.VOTE_TTL_SECONDS)).awaitSingleOrNull()
            logger.debug { "vote_saved chat_id=$chatId approvals=${vote.approvals.size}" }
        } catch (e: RedissonException) {
            logger.error(e) { "vote_save_failed chat_id=$chatId" }
            throw RedisException("Failed to save vote: $chatId", e)
        } catch (e: SerializationException) {
            logger.error(e) { "vote_save_failed chat_id=$chatId" }
            throw RedisException("Failed to save vote: $chatId", e)
        }
    }

    suspend fun approve(
        chatId: String,
        userId: String,
    ): SurrenderVote? {
        val vote = get(chatId) ?: return null
        val updated = vote.approve(userId)
        save(chatId, updated)
        logger.debug {
            "vote_approved chat_id=$chatId user_id=$userId " +
                "approvals=${updated.approvals.size}/${updated.requiredApprovals()}"
        }
        return updated
    }

    suspend fun clear(chatId: String) {
        val key = key(chatId)
        try {
            val bucket = redisson.getBucket<String>(key)
            bucket.delete().awaitSingleOrNull()
            logger.debug { "vote_cleared chat_id=$chatId" }
        } catch (e: RedissonException) {
            logger.error(e) { "vote_clear_failed chat_id=$chatId" }
        }
    }

    private fun key(chatId: String) = "${RedisKeys.SURRENDER_VOTE}:$chatId"

    companion object {
        private val logger = KotlinLogging.logger {}
    }
}

File: game-bot-go/internal/turtlesoup/redis/surrender_vote_store.go
package redis

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"github.com/valkey-io/valkey-go"

	commonvote "github.com/park285/llm-kakao-bots/game-bot-go/internal/common/vote"
	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
)

// SurrenderVoteStore: 게임 항복(Surrender) 투표의 진행 상황(찬성 수, 만료 시간 등)을 Redis에 저장하고 관리하는 저장소
type SurrenderVoteStore struct {
	base   *commonvote.SurrenderVoteStore
	logger *slog.Logger
}

// NewSurrenderVoteStore: 새로운 SurrenderVoteStore 인스턴스를 생성합니다.
func NewSurrenderVoteStore(client valkey.Client, logger *slog.Logger) *SurrenderVoteStore {
	if logger == nil {
		logger = slog.Default()
	}
	return &SurrenderVoteStore{
		base: commonvote.NewSurrenderVoteStore(
			client,
			voteKey,
			time.Duration(tsconfig.RedisVoteTTLSeconds)*time.Second,
		),
		logger: logger,
	}
}

// Get: 현재 활성화된 투표 상태를 조회합니다. 투표가 없으면 nil을 반환합니다.
func (s *SurrenderVoteStore) Get(ctx context.Context, chatID string) (*tsmodel.SurrenderVote, error) {
	vote, err := s.base.Get(ctx, chatID)
	if err != nil {
		return nil, fmt.Errorf("vote get failed: %w", err)
	}
	return vote, nil
}

// Save: 변경된 투표 상태 객체를 Redis에 저장하고 TTL을 갱신합니다.
func (s *SurrenderVoteStore) Save(ctx context.Context, chatID string, vote tsmodel.SurrenderVote) error {
	if err := s.base.Save(ctx, chatID, vote); err != nil {
		return fmt.Errorf("vote save failed: %w", err)
	}
	s.logger.Debug("vote_saved", "chat_id", chatID, "approvals", len(vote.Approvals))
	return nil
}

// Approve: 특정 사용자의 '찬성' 의사를 투표에 반영하고, 갱신된 투표 상태를 반환합니다.
func (s *SurrenderVoteStore) Approve(ctx context.Context, chatID string, userID string) (*tsmodel.SurrenderVote, error) {
	vote, err := s.base.Get(ctx, chatID)
	if err != nil {
		return nil, fmt.Errorf("vote get failed: %w", err)
	}
	if vote == nil {
		return nil, nil
	}

	updated, err := vote.Approve(userID)
	if err != nil {
		return nil, fmt.Errorf("vote approve failed: %w", err)
	}
	if err := s.base.Save(ctx, chatID, updated); err != nil {
		return nil, fmt.Errorf("vote save failed: %w", err)
	}
	s.logger.Debug("vote_approved", "chat_id", chatID, "user_id", userID, "approvals", len(updated.Approvals), "required", updated.RequiredApprovals())
	return &updated, nil
}

// Clear: 투표 상태를 Redis에서 삭제합니다. (투표 완료 또는 취소 시)
func (s *SurrenderVoteStore) Clear(ctx context.Context, chatID string) error {
	if err := s.base.Clear(ctx, chatID); err != nil {
		return fmt.Errorf("vote clear failed: %w", err)
	}
	s.logger.Debug("vote_cleared", "chat_id", chatID)
	return nil
}
